zkevm-circuits/src/evm_circuit/op_execution/pop.rs

```diff
@@ -0,0 +1,305 @@
+
+#[derive(Clone, Debug)]
+struct PopSuccessAllocation<F> {
+    case_selector: Cell<F>,
+}
+
+#[derive(Clone, Debug)]
+pub struct PopGadget<F> {
+    success: PopSuccessAllocation<F>,
+    stack_underflow: Cell<F>, // case selector
+    out_of_gas: (
+        Cell<F>, // case selector
+        Cell<F>, // gas available
+    ),
+}
+
+impl<F: FieldExt> OpGadget<F> for PopGadget<F> {
+    const RESPONSIBLE_OPCODES: &'static [OpcodeId] = &[
+        OpcodeId::POP // 0x50 of op id
+    ];
+
+
+    const CASE_CONFIGS: &'static [CaseConfig] = &[
+        CaseConfig {
+            case: Case::Success,
+            num_word: 0,// no operand required for pop 
+            num_cell: 0, 
+            will_halt: false,
+        },
+        CaseConfig {
+            case: Case::StackUnderflow,
+            num_word: 0,
+            num_cell: 0,
+            will_halt: true,
+        },
+        CaseConfig {
+            case: Case::OutOfGas,
+            num_word: 0,
+            num_cell: 0,
+            will_halt: true,
+        },
+    ];
+
+    fn construct(case_allocations: Vec<CaseAllocation<F>>) -> Self {
+        let [ success, stack_underflow, out_of_gas]: [CaseAllocation<F>; 3] =
+            case_allocations.try_into().unwrap();
+        Self {
+            success: PopSuccessAllocation {
+                case_selector: success.selector.clone(),
+                // no more for pop
+            },
+            stack_underflow: stack_underflow.selector.clone(),
+            out_of_gas: (
+                out_of_gas.selector.clone(),
+                out_of_gas.resumption.unwrap().gas_available.clone(),
+            ),
+        }
+    }
+
+    fn constraints(
+        &self,
+        op_execution_state_curr: &OpExecutionState<F>,
+        op_execution_state_next: &OpExecutionState<F>,
+    ) -> Vec<Constraint<F>> {
+        let pop = Expression::Constant(F::from_u64(80));
+        let OpExecutionState { opcode, .. } = &op_execution_state_curr;
+        let common_polys = vec![(opcode.expr() - pop.clone())];
+        
+        let success = {
+            // interpreter state transition constraints
+            let one = Expression::Constant(F::from_u64(1));
+            let op_execution_state_transition_constraints = vec![
+                op_execution_state_next.global_counter.expr()
+                    - (op_execution_state_curr.global_counter.expr()
+                        + one.clone()),
+                op_execution_state_next.stack_pointer.expr()
+                    - (op_execution_state_curr.stack_pointer.expr()
+                        + one.clone()),
+                op_execution_state_next.program_counter.expr()
+                    - (op_execution_state_curr.program_counter.expr()
+                        + one.clone()),
+                op_execution_state_next.gas_counter.expr()
+                    - (op_execution_state_curr.gas_counter.expr()
+                        + Expression::Constant(F::from_u64(2))),
```

Comment:
nitpick: What do you think we add a constant for `GasQuickStep = 2`

Reply:
nice, I think should place on common module such as  gas_counter.rs , which holds all op codes' gas counter there. it is a good modularization design ? what do u think ?

Reply:
update to latest GasCost.